File: src/libs/task/message.ts
import { listen } from "@tauri-apps/api/event";
import { useTaskStore } from "../../store/task";
import { notifyError, notifyFinish } from "../notification";
import { Running } from "./state_machine";

export const TASK_MESSAGE_EVENT = "transcoding";

export type TaskMessage = TaskMessageRunning | TaskMessageFinished | TaskMessageErrored;

export type TaskMessageRunning = {
  state: "Running";
  id: string;
  progress_type: TaskProgressType;
  raw: string[];
  frame?: number;
  fps?: number;
  bitrate?: number;
  total_size?: number;
  output_time_ms?: number;
  dup_frames?: number;
  drop_frames?: number;
  speed?: number;
};

export type TaskMessageFinished = {
  state: "Finished";
  id: string;
};

export type TaskMessageErrored = {
  state: "Errored";
  id: string;
  reason: string;
};

export type TaskProgressType =
  | TaskProgressTypeUnspecified
  | TaskProgressTypeByDuration
  | TaskProgressTypeByFileSize
  | TaskProgressTypeAuto;

export type TaskProgressTypeByDuration = {
  type: "ByDuration";
  duration: number;
};

export type TaskProgressTypeByFileSize = {
  type: "ByFileSize";
  size: number;
};

export type TaskProgressTypeAuto = {
  type: "Auto";
  duration: number;
  file_size: number;
};

export type TaskProgressTypeUnspecified = {
  type: "Unspecified";
};

/**
 * Starts listening task messages from backend.
 */
listen<TaskMessage>(TASK_MESSAGE_EVENT, async (event) => {
  const { tasks, finishTask, errorTask, updateTask } = useTaskStore.getState();

  const message = event.payload;

  const task = tasks.find((task) => task.id === message.id);
  if (!task) return;

  switch (message.state) {
    case "Running":
      // directly update state for running task
      updateTask(message.id, { state: new Running(message) });
      break;
    case "Errored":
      await errorTask(message.id, message.reason);
      await notifyError(message.reason);
      break;
    case "Finished":
      await finishTask(message.id);
      await notifyFinish(task);
      break;
  }
});

File: src-tauri/src/handlers/tasks/message.rs
pub static TASK_MESSAGE_EVENT: &'static str = "transcoding";

#[derive(Debug, Clone, serde::Serialize)]
pub struct TaskRunningMessage {
    pub id: String,
    pub progress_type: ProgressType,
    pub raw: Vec<String>,
    pub frame: Option<usize>,
    pub fps: Option<f64>,
    pub bitrate: Option<f64>,
    pub total_size: Option<usize>,
    pub output_time_ms: Option<usize>,
    pub dup_frames: Option<usize>,
    pub drop_frames: Option<usize>,
    pub speed: Option<f64>,
}

impl TaskRunningMessage {
    pub fn new(id: String, progress_type: ProgressType) -> Self {
        Self {
            id,
            progress_type,
            raw: Vec::with_capacity(20),
            frame: None,
            fps: None,
            bitrate: None,
            total_size: None,
            output_time_ms: None,
            dup_frames: None,
            drop_frames: None,
            speed: None,
        }
    }

    pub fn clear(&mut self) {
        self.frame = None;
        self.fps = None;
        self.bitrate = None;
        self.total_size = None;
        self.output_time_ms = None;
        self.dup_frames = None;
        self.drop_frames = None;
        self.raw.clear();
    }
}

/// Task message informing task situation.
#[derive(Debug, Clone, serde::Serialize)]
#[serde(tag = "state")]
pub enum TaskMessage<'a> {
    Running(&'a TaskRunningMessage),
    Finished { id: String },
    Errored { id: String, reason: String },
}

impl<'a> TaskMessage<'a> {
    pub fn running(msg: &'a TaskRunningMessage) -> Self {
        Self::Running(msg)
    }

    pub fn finished(id: String) -> Self {
        Self::Finished { id }
    }

    pub fn errored(id: String, reason: String) -> Self {
        Self::Errored { id, reason }
    }
}

#[derive(Debug, Clone, Copy, serde::Serialize)]
#[serde(tag = "type")]
pub enum ProgressType {
    Unknown,
    ByDuration { total: f64 },
    ByFileSize { total: usize },
}
